Add Support for Groups

GCRA101 opened the pull request:
### Issues addressed by this PR

Fixes #509 #500 

<img width="1573" height="846" alt="Test GH Script View" src="https://github.com/user-attachments/assets/65a53fb9-984a-4095-ad9e-acf7615ad9f2" />

<img width="1919" height="1032" alt="Test ETABS Model" src="https://github.com/user-attachments/assets/f03e7d92-65ff-4626-9be8-8fd0a8d9c5a5" />

ETABS Toolkit now allows to push/pull and update objects (i.e. bars, nodes, panels, rigid links and openings) handling corresponding assignments to Groups via Tags.


 ### Test files
Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23513-AddSupportforGroups/[TestScript.gh](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23513-AddSupportforGroups/TestScript.gh?csf=1&web=1&e=NzN4CF)?csf=1&web=1&e=NzN4CF


 ### Changelog
- Add Support for Group in Push/Pull/Update of Bars
- Add Support for Group in Push/Pull/Update of Nodes
- Add Support for Group in Push/Pull/Update of Panels
- Add Support for Group in Push/Pull/Update of Links
- Add Support for Group in Push/Pull/Update of Openings

GCRA101 committed:
Create ETABSGroup.cs class in ETABS_oM

GCRA101 committed:
Delete ETABSGroup.cs

Unnecessary since groupname information is going to be stored in tags.

GCRA101 committed:
Add SetGroup Method to Create/Bar.cs

Add feature allowing to assign pushed bars to corresponding groups.

GCRA101 committed:
Add Support for Group Feature for Push/Pull of Nodes

GCRA101 committed:
Add Support for Group Feature for Push of Bars

GCRA101 committed:
Add Support for Group Feature for Push/Pull of Links

GCRA101 committed:
Add Support for Group Feature for Push/Pull of Panels

GCRA101 committed:
Add Support for Group Feature for Push/Pull of Openings

GCRA101 committed:
Fix Support of Group feature in Push for Links, Openings and Panels

GCRA101 committed:
Add UpdateGroup Feature for Bars

GCRA101 committed:
Add UpdateGroup feature for Nodes

GCRA101 committed:
Add UpdateGroup feature for Panels

GCRA101 committed:
Add UpdateGroup Feature for Links

GCRA101 committed:
Add UpdateGroup Feature for Openings

GCRA101 committed:
Prevent UpdateGroup() method form running with ETABS v16 and v17.

The Method GetGroupAssign() is not supported in any ETABS API earlier than v18.

GCRA101 committed:
Fix syntax error in Update/Bar.cs

GCRA101 committed:
Add missing logic preventing UpdateGroup() method from running with ETABS v16 and v17

GCRA101 commented:
@BHoMBot check installer

michaelhoehn commented:
@BHoMBot check installer

myriamchourfi commented:
Groups pushed and pulled successfully!

samuelbasimalla commented:
Tested this for ETABS Version 22.7.0 using Rhino 7 using the test script provided in @GCRA101's original comment.

The script works as intended:
1. Each object is assigned to different group tags using "Set Property" and pushed to ETABS
2. Groups of objects are uniquely assigned to different group tags using "Set Property" and pushed to ETABS
3. Moreover, the information on group tags is pulled from ETABS along with the BHoM objects and updated as needed and re-pushed using "Update Only" push type.

<img width="601" height="586" alt="image" src="https://github.com/user-attachments/assets/6829e8d8-49f3-415d-a899-72ae9ffd4e9f" />
<img width="683" height="426" alt="image" src="https://github.com/user-attachments/assets/6b4e5b1a-3979-4e34-b9ac-dd3388b837ba" />

GCRA101 committed:
Fix bug panel.Tags.Add() with non perfectly vertical panels

GCRA101 commented:
@BHoMBot check installer

albertbg028 reviewed (dismissed):
New feature works as intended.

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@peterjamesnugent, @Chrisshort92,
fyi, I think this one is ready for merge. Functionality tested by both Myriam and Samuel.
All BHoMBot checks have passed.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Bar.cs:
```diff
@@ -215,6 +215,48 @@ private bool SetObject(Bar bhBar)
+                {
+                    Engine.Base.Compute.RecordError("Could not assign input groups to the bar. Groups not assigned.");
+                    return false;
+                }
```
Do we need a `try catch` - these are generally slow to propagate up.

GCRA101 replied:
@peterjamesnugent,

updated the code in line with your comment above. 
You can find it in latest commit  for your review: https://github.com/BHoM/ETABS_Toolkit/pull/516/commits/d7d3988ea341e754c35cba05f8c3a4b233381458.

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/Bar.cs:
```diff
@@ -215,6 +215,48 @@ private bool SetObject(Bar bhBar)
 
         /***************************************************/
 
+        private bool SetGroup(Bar bhBar)
```
Why can this not be `SetGroup(BHoMObject obj)?

GCRA101 replied:
Peter, 
thanks for the comment.
Fair enough. I will create one single SetGroup(BHoMObject obj) method in _Create.cs and checking that data type of the input within the method itself.

GCRA101 replied:
@peterjamesnugent,

updated the code in line with your comment above. 
You can find it in latest commit  for your review: https://github.com/BHoM/ETABS_Toolkit/pull/516/commits/d7d3988ea341e754c35cba05f8c3a4b233381458.

peterjamesnugent reviewed (dismissed):
A few changes needed to make some methods generic - also how much does this PR conflict with htppotps://github.com/BHoM/ETABS_Toolkit/pull/518 ? One supports unique names whilst the other is supporting tags? I would favour using tags over the unique name PR personally.

GCRA101 committed:
Replace type-specific SetGroup methods with one for generic BHoMObject type

GCRA101 committed:
Replace type-specific ResetGroup methods with one for generic BHoMObject type

peterjamesnugent commented on Etabs_Adapter/CRUD/Create/_Create.cs:
```diff
@@ -109,11 +119,57 @@ private bool CreateObject(IBHoMObject obj)
+            /* 1. CHECK ASSIGNED GROUPS */
 
+            /* Get the list of unique groupNames assigned to the BHoM Bar */
+            List<string> groupNames = obj.Tags.Distinct().ToList();
```
`Tags` are a `Hashset<string>` so you don't need to get `Distinct()` here.

GCRA101 replied:
Thanks for the comment @peterjamesnugent.
I will remove the Distinct() ;)

peterjamesnugent reviewed (commented):
@samuelbasimalla do you want to test this again with the methods being a bit more generic now. 

In the test can you use the folowing workflow:
1. User pushes objects with Tags
2.  User pulls all objects e.g. `Bar`, `Panel` etc.
3. User uses the `FilterRequest in https://github.com/BHoM/BHoM/blob/c7a25f7799782f9025ead7f532b04a6cd33c7ff0/Data_oM/Requests/FilterRequest.cs#L28 to specify the `Tags`
4. User makes user of https://github.com/BHoM/BHoM_Engine/blob/ee49f81ea511d7ee0c66a3c13d0d73f1a3677bd8/Data_Engine/Compute/FilterData.cs#L37 to filter the data as needed.
@GCRA101

samuelbasimalla commented:
> @samuelbasimalla do you want to test this again with the methods being a bit more generic now.
> 
> In the test can you use the folowing workflow:
> 
> 1. User pushes objects with Tags
> 2. User pulls all objects e.g. `Bar`, `Panel` etc.
> 3. User uses the `FilterRequest in https://github.com/BHoM/BHoM/blob/c7a25f7799782f9025ead7f532b04a6cd33c7ff0/Data_oM/Requests/FilterRequest.cs#L28 to specify the `Tags`
> 4. User makes user of https://github.com/BHoM/BHoM_Engine/blob/ee49f81ea511d7ee0c66a3c13d0d73f1a3677bd8/Data_Engine/Compute/FilterData.cs#L37 to filter the data as needed.
>    @GCRA101

Happy to test this as per suggested workflow. 
@GCRA101 is the msi installer available in the SharePoint the latest one, after making the changes?

GCRA101 commented:
@samuelbasimalla,
I'm implementing the latest comment from @peterjamesnugent and run the bhombot so that the new .msi installer can be generated. I'll let you know as soon as it is done. ;)

GCRA101 committed:
Remove Distinct() from groupNames assignment

GCRA101 commented:
@peterjamesnugent, removed Distinct() method as suggested.
Updated code in latest commit https://github.com/BHoM/ETABS_Toolkit/pull/516/commits/aa299a2011ab89c650d8b11f7304baf1642c8b49

GCRA101 commented:
@BHoMBot check installer

GCRA101 commented:
@peterjamesnugent, @samuelbasimalla, @Chrisshort92,
latest msi installer to be used for final testing prior to merge stored on sharepoint at following location:  https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23516-AddSupportForGroups/ETABS_Toolkit-%23509-AddSupportForGroups.msi?csf=1&web=1&e=5Ygy1S

samuelbasimalla commented:
> 3\. https://github.com/BHoM/BHoM/blob/c7a25f7799782f9025ead7f532b04a6cd33c7ff0/Data_oM/Requests/FilterRequest.cs

@GCRA101 Tested as per suggested workflow.
1. Pushed objects as per test script with the tags --> **Bars are not being assigned tags**. All other objects behaving as expected.
2. Used FilterRequest component and specified tags to pull all the objects back ---> Working as expected for all elements that are being assigned tags in step 1 above.

I am not sure about the FilterData component how it can be used to filter the data further. Is there a sample script from other toolkits for this?

GCRA101 committed:
Fix mistake group assigment to nodes/bars.

GCRA101 commented:
@samuelbasimalla,
thanks again for your checks. 
My bad, did a little mistake while reshuffling the code in the latest updates.
Fixed the issue and now all working fine. Tested Push, Pull and Update and it works for all objects including the bars.

@peterjamesnugent,
I'm happy to merge the code as per latest commit https://github.com/BHoM/ETABS_Toolkit/pull/516/commits/d5133e71f814717aebc3afb3ab5a2179c8f4cf6d

GCRA101 commented:
@BHoMBot check installer

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 committed:
Merge branch 'develop' into ETABS_Toolkit-#509-AddSupportForGroups

GCRA101 commented:
@BHoMBot check installer

GCRA101 commented:
@BHoMBot fix copyright headers ref. 59639943524

BHoMBot committed:
Resolve copyright compliance

GCRA101 commented:
@BHoMBot check installer

GCRA101 commented:
@BHoMBot check ready-to-merge

albertbg028 reviewed (approved):
Feature has been tested and behavior was as expected.

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

GCRA101 commented:
@BHoMBot check ready-to-merge

peterjamesnugent reviewed (approved):
All changesI suggested have been made - I have added https://github.com/BHoM/ETABS_Toolkit/issues/523 to make sure the `SetGroup` method is easy to find and users won't necessarily look at the bottom of the _Create method.

samuelbasimalla commented:
@GCRA101 @peterjamesnugent 
I was having issues pushing groups other day. The tags do get assigned when I SetProperty but when I tried to push it back to ETABS, it fails.
I tried going back to the test script in the folder to test it again. I got the following error below. The same test script worked earlier when I tested it before PR got merged.
<img width="518" height="208" alt="image" src="https://github.com/user-attachments/assets/aaea5db2-a863-437a-9ac9-bab6bcf4854b" />

GCRA101 commented:
@samuelbasimalla, @peterjamesnugent 
thanks for reaching out on the above.
I've retested the groups creation using the test script and it all works fine from my end (see below).

<img width="1918" height="1031" alt="image" src="https://github.com/user-attachments/assets/76016112-b19d-4d11-a122-e3312d336608" />



Is the issue still occurring on your machine ?

samuelbasimalla commented:
@GCRA101 I tested this now using 9.1 beta and it's all good. Thanks a lot. I could have used a wrong installer last time.
